### ISSUE TITLE
Drag never "ends"

### ISSUE THREAD

**Issue by ssilve1989** [react-dnd/react-dnd#557]
Sometimes when attempting to drag a component, things seem to get stuck at beginDrag, which in turn persists isDragging as true.

Essentially the problem is thus, when dragging, `display : none` is set to my component and the browser drag preview takes over, on drop the position of the component is updated. However sometimes, under seemingly irreproducible conditions, as soon as you begin to drag you
- Get no browser preview
- isDragging remains true after release the mouse
- in turn, display : none is still set because isDragging is still true

Its also interesting to note that i placed a `hover()` in the DropTarget to see if when this happened it was "hovering" and it wasn't. It just seems to completely get stuck at beginDrag with 0 errors in the console. I did the same for endDrag and endDrag was never called. drop() does not get called either when this happens. The only method that gets called is beginDrag()

DragSource

```
{
    beginDrag(props) {
        const { id, left, top } = props;
        return { id, left, top };
    }
}
```

DropTarget

```
{
drop(props, monitor, component) {
        const item  = monitor.getItem();
        const delta = monitor.getDifferenceFromInitialOffset();
        const left  = Math.round(item.left + delta.x);
        const top   = Math.round(item.top + delta.y);

        component.moveBox(item.id, left, top);
    }
}
```

**Comment by ssilve1989** [comment]
Update: When this happens, if i go into the DOM and toggle off the `display: none` and attempt to drag it agan, you can tell it still thinks its dragging since you get the error
`browser.js?9520:40 Uncaught Invariant Violation: Cannot call beginDrag while dragging.`

**Comment by puneetbhakar** [comment]
I'm having the similar issue. I checked the actions through Redux DevTools chrome extension and found out that only `BEGIN_DRAG` and `PUBLISH_DRAG_SOURCE` actions were fired i.e neither `HOVER` nor `END_DRAG` were fired. I also added console logs in `dnd-core` and `react-dnd-html5-backend` and found that` endDragIfSourceWasRemovedFromDOM() `was called even though the source the was not removed from the DOM. You can see the code [here](https://github.com/puneetbhakar/almabaseReact/blob/master/src/components/Random.js). 
Any help would be appreciated.

Thanks!

**Comment by ssilve1989** [comment]
bump

**Comment by ncphillips** [comment]
Super bump. For some reason this wasn't an issue, but suddenly became an issue. This is breaking production features for me. I'm going to try and figure out a solution, but it would be greatly appreciated if @gaearon or someone else with strong understanding of this library could chip in.

**Comment by TheAncientGoat** [comment]
This was happening to me - until I found a subtle error in code I wrote in the endDrag event; fixing my error resolved it, but it might be better to ensure the drag state gets reset no matter what happens in the user-provided code?

**Comment by rroslaniec** [comment]
@up I also removed that error moving logic from drop (DropTarget) to endDrag (DropSource). What is the difference?

**Comment by siemiatj** [comment]
Same here. Maybe it's the same thing as https://github.com/react-dnd/react-dnd/issues/455 but in my case I started using `v4()` for keys and that didn't help. So I only kept the drop method to check if we're not dropping on the same element + I'm changing the indexes on monitor. The actual drop happens in `endDrop` : 

```
/**
 * DragSource specification object
 */
const tileSource = {
  beginDrag(props) {
    return {
      item: props.item,
      index: props.index,
    };
  },

  endDrag(props, monitor) {
    const dropIndex = monitor.getItem().index;
    const dragIndex = props.index;

    props.moveItem(dragIndex, dropIndex);
  }
};

/**
 * DropTarget specification object
 */
const tileTarget = {
  drop(props, monitor) {
    const dragIndex = monitor.getItem().index;
    const dropIndex = props.index;

    // Don't replace items with themselves
    if (dragIndex === dropIndex) {
      return undefined;
    }

    monitor.getItem().index = dropIndex;

    return undefined;
  }
};
```

**Comment by Zertz** [comment]
This error typically happens when you use the `drop` method to handle changing state, rather than just returning an object.

```js
const dragSource = {
  beginDrag (props) {
    return {
      hello: props.hello
    }
  },
  endDrag (props, monitor) {
    const drag = monitor.getItem().hello
    const drop = monitor.getDropResult().hello

    props.onDragAndDropEnded(drag, drop)
  }
}

const dropTarget = {
  drop (props, monitor) {
    return {
      hello: props.hello
    }
  }
}
```

**Comment by nvanselow** [comment]
If anyone is still having this issue and needs a quick fix: 
I did not have this issue when I moved elements during `hover()` using the Touch Backend (https://github.com/yahoo/react-dnd-touch-backend).

I was able to turn on and off the issue by switching the backend. If I have a chance, I will further investigate the difference between the two backends to see if I can create a PR for the Html5Backend.

`export default DragDropContext(TouchBackend({enableMouseEvents: true, delay: 150}))(DragDropContainer);`
